# This diff contains the following commits:
#
# cb2a389 Execute sub-Makefiles using make recursion.
# a62df9c Avoid SystemError when running the yapps2.py script.
# 4ad81ae Allow override of Python that runs yapps2.py.

diff --git a/Makefile b/Makefile
index 8f364d2..f8ef6b6 100644
--- a/Makefile
+++ b/Makefile
@@ -9,11 +9,10 @@ package: setup.py sources drel
 #	python setup.py bdist_wininst
 #
 documentation:
-	cd docs;make
+	$(MAKE) -C docs
 #
 sources:
-	cd pycifrw;make
+	$(MAKE) -C pycifrw
 #
 drel:
-	cd pycifrw/drel;make
-#
+	$(MAKE) -C pycifrw/drel
diff --git a/pycifrw/Makefile b/pycifrw/Makefile
index 2f44218..809c777 100644
--- a/pycifrw/Makefile
+++ b/pycifrw/Makefile
@@ -1,6 +1,11 @@
 # Makefile for ANBF Python based Cif handling modules
 
+# Customizable variables:
 #
+# Instance of Python to be used for the yapps2.py script.
+PYTHON = python3
+
+
 package: CifFile.py StarFile.py Parsers documentation
 #	python setup.py bdist_wininst
 #
@@ -15,32 +20,36 @@ documentation: CifFile.nw YappsStarParser.nw StarFile.nw
 	noweave -html -index -filter l2h StarFile.nw > StarFile.html
 	noweave -html -index -filter l2h YappsStarParser.nw > YappsStarParser.html
 	noweave -html -index -filter l2h TypeContentsParser.nw > TypeContentsParser.html
-# 
+#
 Parsers: YappsStarParser_STAR2.py YappsStarParser_1_1.py YappsStarParser_1_0.py  TypeContentsParser.py \
 	YappsStarParser_2_0.py
-	
+
 #
-clean: 
+clean:
 	rm -f *.pyc *.g
+
+# Local helper variables:
 #
-# Note the pythonpath variable as yapps requires its runtime to run
+# Command to execute the yapps2.py script.  Note it sets the PYTHONPATH
+# environment variable as yapps requires its runtime to run.
+YAPPS2CMD = PYTHONPATH=. $(PYTHON) ./yapps3/yapps2.py
+
 YappsStarParser_1_0.py: YappsStarParser.nw
 	notangle -R1.0_syntax YappsStarParser.nw > YappsStarParser_1_0.g
-	export PYTHONPATH=.;python3 ./yapps3/yapps2.py YappsStarParser_1_0.g
+	$(YAPPS2CMD) YappsStarParser_1_0.g
 #
 YappsStarParser_1_1.py: YappsStarParser.nw
 	notangle -R1.1_syntax YappsStarParser.nw > YappsStarParser_1_1.g
-	export PYTHONPATH=.;python3 ./yapps3/yapps2.py YappsStarParser_1_1.g
+	$(YAPPS2CMD) YappsStarParser_1_1.g
 #
 YappsStarParser_2_0.py: YappsStarParser.nw
 	notangle -RCIF2_syntax YappsStarParser.nw > YappsStarParser_2_0.g
-	export PYTHONPATH=.; python3 ./yapps3/yapps2.py YappsStarParser_2_0.g
+	$(YAPPS2CMD) YappsStarParser_2_0.g
 #
 YappsStarParser_STAR2.py: YappsStarParser.nw
 	notangle -RSTAR2_syntax YappsStarParser.nw > YappsStarParser_STAR2.g
-	export PYTHONPATH=.; python3 ./yapps3/yapps2.py YappsStarParser_STAR2.g
+	$(YAPPS2CMD) YappsStarParser_STAR2.g
 #
 TypeContentsParser.py: TypeContentsParser.nw
 	notangle -RTypeContents_syntax TypeContentsParser.nw > TypeContentsParser.g
-	export PYTHONPATH=.; python3 ./yapps3/yapps2.py TypeContentsParser.g
-
+	$(YAPPS2CMD) TypeContentsParser.g
diff --git a/pycifrw/yapps3_compiled_rt.py b/pycifrw/yapps3_compiled_rt.py
index d3efd69..f4a040c 100644
--- a/pycifrw/yapps3_compiled_rt.py
+++ b/pycifrw/yapps3_compiled_rt.py
@@ -40,12 +40,22 @@ keeps track of the parse stack.
 # grammar to make a standalone module.
 
 import sys, re
-try:
-    from CifFile import StarScan
+
+
+# For normal installation this module is "CifFile.yapps3_compiled_rt"
+# and StarScan is an extension module within the parent CifFile module.
+if __name__.startswith('CifFile.'):
+    from . import StarScan
     have_star_scan = True
-except ImportError:
+# Otherwise assume this is imported from the yapps3/yapps2.py script
+# that is executed from Makefile to generate YappsStarParser sources.
+else:
+    assert __name__ == 'yapps3_compiled_rt', "Unexpected module name."
+    assert sys.argv[0].endswith('yapps2.py'), (
+        "This should be reached only when running yapps2.py in Makefile.")
     have_star_scan = False
 
+
 class SyntaxError(Exception):
     """When we run into an unexpected token, this is the exception to use"""
     def __init__(self, charpos=-1, msg="Bad Token", context=None):
